src/main/java/org/springframework/retry/annotation/AnnotationAwareRetryOperationsInterceptor.java

```diff
@@ -357,6 +361,9 @@ private RetryPolicy getRetryPolicy(Annotation retryable) {
 		for (Class<? extends Throwable> type : excludes) {
 			policyMap.put(type, false);
 		}
+		for (Class<? extends Throwable> type : rethrowExceptions) {
+			policyMap.put(type, false);
+		}
```

Comment:
it is using `rethrowExceptions`, the previous `for` is using `excludes` to iterate over.

Reply:
Right but why do we need `rethrowExceptions` when we already have `excludes`?

Reply:
Using `rethrowExceptions` on `Retryable` sets the `noRecoveryForNotRetryable` flag of RetryTemplate, so the listed exceptions are not only not-retryable, but also directly rethrown (without recovery and wrapping).

Reply:
So, just set `rethrow` alongside `excludes`.

Reply:
You're right @garyrussell , in this implementation (second commit), excludes + rethrow is the same. But it was not my original intention, it was a mistake I did in the second commit - sorry!

There is a difference between the behavior of the first and second commit: **the recovery method**.

What I need is to skip the recovery method only for the listed exceptions, but `rethrow=true` means the recovery method will be skipped for all.

```java
@Retryable(include = Ex1.class, excludes = Ex2.class, rethrow = true)
public Object doSomething() { throw new Ex3(); }

@Recover Object recovery() { return "x"; }  

...
srv.doSomething();  // throws Ex3
```
vs
```java
@Retryable(include = Ex1.class, rethrowExceptions = Ex2.class)
public Object doSomething() { throw new Ex3(); }

@Recover Object recovery() { return "x"; }

...
srv.doSomething();  // returns "x"
```

I guess RetryPolicy cannot be used for this as it is exception handling rather than retrying...

Does it make sense?

Reply:
I have reverted the second commit and added tests to demonstrate this behavior: `testRethrowForAnotherNotRetryable` and `testRethrowExceptionsForAnotherNotRetryable`

https://github.com/spring-projects/spring-retry/pull/281/files#diff-a6717c7c5ec799b50bdfde5dada472a3cd9277b3556c9136fb9292770b63c28cR436

Reply:
>means the recovery method will be skipped for all.

You mean for all not-retryable exceptions, right? Retryable exceptions should always go to the recovery method, even with `rethrow=true`.

So, if I understand you correctly, you want `include` exceptions to be retried, `rethrowable` exceptions to be rethrown without modification, skipping any matching recovery method, and any not listed exceptions to go straight to the recovery method, if one matches, and rethrown without modification if no recovery method matches?

If so, we would have to clearly document that in the README and javadocs.

I really think the `@Recover` annotation is complicated enough already but; I still think we should try to achieve your use case some other way. Need to think some more.

I think this will have to miss tomorrow's 1.3.3 release.

Reply:
Yes, all not-retryable exceptions. Exactly as you described it, @garyrussell .

The use-case is following: 

There are some exceptions I want to retry (`include`), things such as timeouts and so, and then recovery when even retrying fails. Then there are exceptions I don't want to do anything with (`rethrowExceptions`), such as authorization exceptions, and rethrow them directly to the caller. For all the others I do want to recover.

Putting this logic into the recover method means a lot of additional code, that's why I decided to open this PR.

We can keep this open until the decision is made.

Many thanks!

Reply:
I wonder if this would be easier for users to understand...

```java
@Retryable(include = { ... }, exclude = { ... }, rethrow = true, recoverAllOthers = true)
```
If `rethrow` is true and `recoverAllOthers` is false (default), all excluded + not included exceptions are rethrown immediately.

If `rethrow` is true and `recoverAllOthers` is true, all excluded exceptions are rethrown immediately, all others (not in `include` or `exclude` and not a subclass of any of those) go straight to the recover method (if one matches, or rethrown otherwise).

This would probably need changes to the classifier - in another project I subclassed it and made `getClassified()` public so I could check if a particular exception is explicitly classified; I see no reason we can't make it public (or at least a version that returns an unmodifiable map).

I am not married to the name `recoverAllOthers`, I am just concerned about adding a third type of "exclude" classes; we already have two - `exclude` and not `include`.

Reply:
I am not quite happy about this for several reasons:

1) `recoverAllOthers` does really make sense only in a combination with `rethrow`.  This makes it more complex and hard to reason about. The default value is false, but the behavior without `rethrow=true` is the same or at least very similar to being set to true (not-retryable go to the recover method) - this is confusing.

2) It is giving a second responsibility to `exclude` which makes it more complex. So far I used to think about `includes` and `excludes` as about settings of what is retryable and what is not. Now, there is a new responsibility of what is to rethrow/recover. That makes particular valid use-cases impossible as far as I can see:

Valid use-case: I want to retry all exceptions `A` (and subtypes) except its subtype `AB`. Further, I want to immediately rethrow (no recovery) exceptions of type `B`. For all others I expect the recover method being triggered:

```java
@Retryable(include = A.class, excludes = AB.class, rethrowExceptions = B.class)
public Object exA() { throw new A(); }
public Object exAA() { throw new AA(); }
public Object exAB() { throw new AB(); }
public Object exB() { throw new B(); }
public Object exC() { throw new C(); }

@Recover Object recover() { return "x"; } 

srv.exA();  // "x", retries, then recover
srv.exAA();  // "x", retries, then recover
srv.exAB();  // "x", no retries, but recover
srv.exB();  // B!!!, no retries, no recover
srv.exC();  // "x", no retries, but recover
````

I have added two tests to demonstrate this: `testRethrowExceptionsForAnotherNotRetryableWithExcludes` and `testRethrowExceptionsForAnotherNotRetryableWithExcludesRecovered`

https://github.com/spring-projects/spring-retry/pull/281/files#diff-a6717c7c5ec799b50bdfde5dada472a3cd9277b3556c9136fb9292770b63c28cR487

3) `rethrowExceptions` is in fact just a specialized variant of `rethrow` which is quite easy to understand:

- `rethrow`: rethrow with no recover *all* not-retryable exceptions
- `rethrowExceptions`: rethrow with no recover *these* not-retryable exceptions only

So, we have `includes` and `excludes` for telling which exceptions are retryable, and we have `rethrow` and `rethrowExceptions` to tell which not-retryable exceptions are to rethrow without recover. Two quite separate topics. Natural to reason about.

Does it make sense, @garyrussell ?

Reply:
I see what you are saying, but I still think it will be confusing for the casual user so clear documentation is going to be crucial.

@artembilan Do you have an opinion?

Reply:
How about to make that `rethrow` not as plain `boolean`, but `enum` with possible logic behind. Right now it look like 3 scenarios.
So, no need in that extra `recoverAllOthers`

Reply:
That won't work; he has 3 categories of exceptions:

- retryable and recoverable
- not retryable but recoverable
- not retryable and not recoverable (rethrow)

I suppose we could add a fourth category

- retryable and not recoverable (rethrow if the last retry fails and don't recover, even if a recover method matches).

The challenge, IMO, is to make these configurable in an intuitive manner; I am just not comfortable with `include`, `exclude`, and `rethrowExceptions`.

Maybe we need 3 (or 4) new attributes that are (as a collection) mutually exclusive with `include` and `exclude`.

Reply:
... and get rid of that confusing `rethrow`, I guess...

Reply:
Sine 1.3.3 hasn't been built yet, maybe we should revert that commit until this is sorted out?

Reply:
That's sounds reasonable.
Looks like that one was requested by @ttulka as well: https://github.com/spring-projects/spring-retry/commit/ec245fd1bac5c392c091757a47863f2769c3e21b.

So, we are safe so to not release anything on the matter and continue discussion 😄

Reply:
Done - maybe we should move this whole thing to the new 2.0 branch where we can overhaul the entire attribute list?

Reply:
Not exactly happy about not getting anything in the next release, but as far as you guys decide on the API attributes,  I would love to implement it if you want me to.

All in all, this feature is going to save me a lot of work!

Many thanks

Reply:
I understand you are "not happy", but you realized that the previous commit did not satisfy all your needs and submitted this new PR, that does not use the new attribute from the previous commit.

It made no sense to us to release a version with a new attribute that would immediately become obsolete in the "next" release.

We **are happy** that you are willing to contribute, but we need to finalize a design that makes sense for the entire community, and provides minimal confusion for users with simpler use cases.

Thanks, again, for your contribution; we will continue to work this through.